File: python3/tree_node.py
import unittest

# Definition for a binary tree node
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        self.next = None

    @staticmethod
    def treeNodeToString(root) -> str:
        if not root:
            return "[]"
        output = ""
        queue = [root]
        current = 0
        while current != len(queue):
            node = queue[current]
            current = current + 1

            if not node:
                output += "null, "
                continue

            output += str(node.val) + ", "
            queue.append(node.left)
            queue.append(node.right)
        return "[" + output[:-2] + "]"

    @staticmethod
    def stringToTreeNode(input: str):
        input = input.strip()
        input = input[1:-1]
        if not input:
            return TreeNode()

        inputValues = [s.strip() for s in input.split(',')]
        root = TreeNode(int(inputValues[0]))
        nodeQueue = [root]
        front = 0
        index = 1
        while index < len(inputValues):
            node = nodeQueue[front]
            front = front + 1

            item = inputValues[index]
            index = index + 1
            if item != "null":
                leftNumber = int(item)
                node.left = TreeNode(leftNumber)
                nodeQueue.append(node.left)

            if index >= len(inputValues):
                break

            item = inputValues[index]
            index = index + 1
            if item != "null":
                rightNumber = int(item)
                node.right = TreeNode(rightNumber)
                nodeQueue.append(node.right)
        return root

    def __eq__(self, other):
        if other != None:
            if self.val == other.val:
                return self.right == other.right and self.left == other.left
        return False

class TestSolution(unittest.TestCase):
    def test_eq1(self):
        t1 = TreeNode.stringToTreeNode('[2,1,3]')
        t2 = TreeNode.stringToTreeNode('[2,1,3]')
        self.assertEqual(t1 == t2, True)

    def test_eq2(self):
        t1 = TreeNode.stringToTreeNode('[2,1,3]')
        t2 = TreeNode.stringToTreeNode('[1,2,3]')
        self.assertEqual(t1 == t2, False)

if __name__ == '__main__':
    unittest.main()

File: go_legacy/utils/tree_node.go
package utils

import (
	"strconv"
	"strings"
)

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func StringToTreeNode(s string) *TreeNode {
	s = s[1:len(s)-1]
	sSplit := strings.Split(s, ",")
	if len(sSplit) == 0 || sSplit[0] == "" {
		return nil
	}
	val, err := strconv.Atoi(sSplit[0])
	if err != nil {
		panic(err)
	}
	root := &TreeNode{
		Val: val,
	}
	queue := []*TreeNode{root}
	i := 1
	for i < len(sSplit) {
		child, err := strconv.Atoi(sSplit[i])
		if err == nil {
			left := &TreeNode {
				Val: child,
			}
			queue[0].Left = left
			queue = append(queue, left)
		}
		i++
		if i >= len(sSplit) {
			break
		}
		child, err = strconv.Atoi(sSplit[i])
		if err == nil {
			right := &TreeNode{
				Val: child,
			}
			queue[0].Right = right
			queue = append(queue, right)
		}
		i++
		queue = queue[1:]
	}
	return root	
}

func TreeNodeToString(t *TreeNode) string {
	s := "["
	output := ""
	if t != nil {
		q := []*TreeNode{t}
		for len(q) != 0 {
			if q[0] == nil {
				output += "null,"
				continue
			} else {
				output += strconv.Itoa(q[0].Val) + ","
			}
			q = append(q, q[0].Left, q[0].Right)
		}
	}
	if output != "" {
		output = output[:len(output)-1]
	}
	s += output + "]"
	return s
}

func TreeNodeEquals(t1 *TreeNode, t2 *TreeNode) bool {
	if t1 == nil && t2 == nil {
		return true
	}
	if (t1 == nil && t2 != nil) || (t1 != nil && t2 == nil) {
		return false
	}
	if t1.Val != t2.Val {
		return false
	}
	return TreeNodeEquals(t1.Left, t2.Left) && TreeNodeEquals(t1.Right, t2.Right)
}
